src/UglyToad.PdfPig/Function/PdfFunction.cs

```diff
@@ -0,0 +1,224 @@
+        public PdfRange GetDomainForInput(int n)
+        {
+            return new PdfRange(Domain, n);
+        }
+
+        /// <summary>
+        /// (Required) The function type:
+        /// <para>0 Sampled function</para>
+        /// <para>2 Exponential interpolation function</para>
+        /// <para>3 Stitching function</para>
+        /// <para>4 PostScript calculator function</para>
+        /// </summary>
+        public abstract int FunctionType { get; }
+
+        /// <summary>
+        /// PdfFunction
+        /// </summary>
+        /// <param name="functionDictionary"></param>
+        /// <param name="pdfTokenScanner"></param>
+        public PdfFunction(DictionaryToken functionDictionary, IPdfTokenScanner pdfTokenScanner)
+        {
+            numberOfOutputValues = -1;
+            numberOfInputValues = -1;
+
+            if (functionDictionary != null)
+            {
+                FunctionDictionary = functionDictionary;
+                if (functionDictionary.TryGet<ArrayToken>(NameToken.Domain, pdfTokenScanner, out var domain))
+                {
+                    Domain = domain;
+                }
+                else
+                {
+                    throw new ArgumentException("Domain is Required.");
+                }
+
+                if (functionDictionary.TryGet<ArrayToken>(NameToken.Range, pdfTokenScanner, out var range))
+                {
+                    Range = range;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Clip the given input values to the ranges.
+        /// </summary>
+        /// <param name="inputValues">the input values</param>
+        /// <returns>the clipped values</returns>
+        protected float[] clipToRange(float[] inputValues)
+        {
+            ArrayToken rangesArray = Range;
+            float[] result;
+            if (rangesArray?.Length > 0)
+            {
+                float[] rangeValues = rangesArray.Data.Select(x => (float)((NumericToken)x).Double).ToArray();
+                int numberOfRanges = rangeValues.Length / 2;
+                result = new float[numberOfRanges];
+                for (int i = 0; i < numberOfRanges; i++)
+                {
+                    int index = i << 1;
+                    result[i] = clipToRange(inputValues[i], rangeValues[index], rangeValues[index + 1]);
+                }
+            }
+            else
+            {
+                result = inputValues;
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Clip the given input value to the given range.
+        /// </summary>
+        /// <param name="x">x the input value</param>
+        /// <param name="rangeMin">rangeMin the min value of the range</param>
+        /// <param name="rangeMax">rangeMax the max value of the range</param>
+        /// <returns>the clipped value</returns>
+        protected float clipToRange(float x, float rangeMin, float rangeMax)
```

Comment:
I'd prefer to avoid Java style naming even for private/protected methods, so `ClipToRange` and `Interpolate` would be more in keeping with the existing code style.

Reply:
agreed